### PR TITLE
GCE configure and deploy script

### PR DESCRIPTION
**Pull request by eap**
This change adds an auto-deploy script and small modifications to Leonardo's code to enable that script. As for now, it is a work in progress; the connection between launched Notebook servers and users are not authenticating properly. 

* automation/gce-deploy deploy.py
  * Script to configure and launch Leo on a "bare" GCP project. Can be run with minimal flags for a deploy with nearly full automation, or a number of flags can be added to re-use existing infrastructure like running Cloud SQL instances.
<br>

* automation/gce-deploy/README.md
  * Quickstart and detailed instructions for deploy.py.
<br>

* /src/.../leonardo/auth/SimpleServiceAccountProvider.scala
  * Service account provider with no SAM dependency.

Before submission:
- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SimpleServiceAccountProvider.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+import com.typesafe.config.Config
+import org.broadinstitute.dsde.workbench.leonardo.model.ServiceAccountProvider
+import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
+import org.broadinstitute.dsde.workbench.model.google.GoogleProject
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class SimpleServiceAccountProvider(val config: Config) extends ServiceAccountProvider(config) {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SimpleServiceAccountProvider.scala]: Just noting this is the exact same implementation as `DefaultServiceAccountProvider`:

https://github.com/DataBiosphere/leonardo/blob/develop/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/DefaultServiceAccountProvider.scala

You could probably just use that one.

File: src/main/resources/jupyter/init-actions.sh [hunk 2]
```diff
@@ -92,6 +92,14 @@ if [[ "${ROLE}" == 'Master' ]]; then
       echo "" > /etc/google_application_credentials.env
     fi
 
+    # Pre-load docker images using gcloud. This prevents auth errors while fetching a non-public GCR images.
```
Comment (rtitle) [inline src/main/resources/jupyter/init-actions.sh]: Neat. I was testing GCR earlier using
```
gcloud docker --authorize-only
```
but this should work too I think.

**Comment by rtitle** [comment]
Cool! This looks great. I don't really have any specific comments on the `deploy.py` code - it looks great to me and if it works then :shipit:.

Just curious, did you consider using something like Cloud Deployment Manager? I haven't looked too closely, so I'm not sure it could support all the things needed by Leo.

Thanks also for making GCR images work in `init-actions.sh`. I think I'm going to push for using GCR internally here as well, at least for the `leonardo-notebooks` image -- I feel like it should make cluster creation faster & more reliable.

Lastly I'm just running our integration tests against these changes (I don't think they should break anything). I am not sure why our int-tests don't kick off automatically for external PRs.

**Comment by eap** [comment]
Regarding Deployment Manager; I've used it before and I find it better suited for cases where a entire system is deployed end-to-end every time with minimal case-by-case changes. With this script, I wanted the option to re-deploy just the VM if needed. That's possible with deployment manager but much harder to set up. If Leonardo had multiple service layers or required something a load-balancer it might make sense to migrate to deployment manager, but a lot of the config logic here would still still be needed to prep a template deployment config. See [forseti-security](http://github.com/GoogleCloudPlatform/forseti-security) for an example of a project that combines a config script with a deployment template.

As to your other comments, I'll be adding a commit to address them. Both are good catches.

**Comment by rtitle** [comment]
Thanks for that context @eap -- makes sense. Just waiting for that extra commit.

**Pull request merged**